File: app/src/main/java/com/peoplesafe/views/WiperSwitch.java
package com.peoplesafe.views;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Matrix;
import android.graphics.Paint;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;

import com.monk.aidldemo.R;
import com.monk.commonutils.LogUtil;


/**
 * @author MonkHank
 * @date 2018-05-10.
 */
public class WiperSwitch extends View implements View.OnTouchListener {

    private Bitmap bg_on, bg_pause, bg_off, slipper_pause, slipper_white;
    // 当前滑块x
    /**
     * 0：bg_on.getWidth() - slipper_pause.getWidth()
     * 1：0
     * 2：bg_on.getWidth() / 3
     */
    private int nowX;

    // 0:on  1:pause  2:off
    private int nowStatus = -1;
    private final int ON=0;
    private final int PAUSE=1;
    private final int OFF=2;

    private Matrix matrix;
    private Paint paint;

    private OnChangedListener listener;

    public WiperSwitch(Context context) {
        super(context);
        init();
    }

    public WiperSwitch(Context context, AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    private void init() {
        bg_off = BitmapFactory.decodeResource(getResources(), R.drawable.slide_switch_off);
        bg_on = BitmapFactory.decodeResource(getResources(), R.drawable.slide_switch_on);
        bg_pause = BitmapFactory.decodeResource(getResources(), R.drawable.slide_switch_stop);
        slipper_pause = BitmapFactory.decodeResource(getResources(), R.drawable.slide_dot);
        slipper_white = BitmapFactory.decodeResource(getResources(), R.drawable.slide_switch_block_30x90);
        setOnTouchListener(this);
        matrix = new Matrix();
        paint = new Paint();
        nowX = bg_on.getWidth() / 3;
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        if (nowX < (bg_on.getWidth() / 3)) {
            canvas.drawBitmap(bg_off, matrix, paint);//画出关闭时候的背景
        } else if (nowX > (bg_on.getWidth() / 3*2)) {
            canvas.drawBitmap(bg_on, matrix, paint);//画出打开时候的背景
        } else {
            if (nowStatus == ON) {
                canvas.drawBitmap(bg_on, matrix, paint);
            } else if (nowStatus == OFF) {
                canvas.drawBitmap(bg_off, matrix, paint);
            } else if (nowStatus == PAUSE) {
                canvas.drawBitmap(bg_pause, matrix, paint);//画出暂停时候的背景
            }
        }
        // 对滑块滑动进行异常处理，不能让滑块出界
        if (nowX < 0) {
            nowX = 0;
        } else if (nowX > bg_on.getWidth() - slipper_pause.getWidth()) {
            nowX = bg_on.getWidth() - slipper_pause.getWidth();
        }
        // 画出滑块
        LogUtil.v("WiperSwitch_onDraw", "nowX:" + nowX+"\tbg_on.getWidth():"+bg_on.getWidth()
                +"\tslipper_pause.getWidth():"+ slipper_pause.getWidth());
        switch (nowStatus) {
            case -1:
                canvas.drawBitmap(bg_pause, matrix, paint);//画出暂停时候的背景
                canvas.drawBitmap(slipper_white, bg_off.getWidth() / 3, 0, paint);
                break;
            case ON:
            case OFF:
                canvas.drawBitmap(slipper_white, nowX, 0, paint);
                break;
            case PAUSE:
                canvas.drawBitmap(slipper_pause, bg_off.getWidth() / 3, 0, paint);
                break;
        }
        if (action == MotionEvent.ACTION_UP) {
            if (nowX < bg_on.getWidth() / 3) {
                nowX += 10;
                if (nowX > bg_on.getWidth() / 3)
                    nowX = bg_on.getWidth() / 3;
                postInvalidateDelayed(10);
            } else if (nowX > bg_on.getWidth() / 3) {
                nowX -= 10;
                if (nowX < bg_on.getWidth() / 3)
                    nowX = bg_on.getWidth() / 3;
                postInvalidateDelayed(10);
            }
        }
    }

    private int action;
    @Override
    public boolean onTouch(View v, MotionEvent event) {
        getParent().requestDisallowInterceptTouchEvent(true);
        action = event.getAction();
        float x = event.getX();
        float y = event.getY();
        switch (action) {
            case MotionEvent.ACTION_DOWN:
                if (x > bg_on.getWidth() || y > bg_on.getHeight()) {
                    return false;
                } else {
                    nowX = (int) x;
                }
                LogUtil.e("ACTION_UP","downX:"+x);
                obtainStatus(x);
                break;
            case MotionEvent.ACTION_MOVE:
                nowX = (int) x;
                LogUtil.i("ACTION_UP","moveX:"+x);
                obtainStatus(x);
                break;
            case MotionEvent.ACTION_UP:
                LogUtil.v("ACTION_UP", "nowStatus:" + nowStatus);
                if (listener != null)
                    listener.onChanged(this, nowStatus);
                break;
        }
        invalidate();

        return true;
    }

    private void obtainStatus(float x) {
        if (x >= bg_on.getWidth() / 3 *2) {// on 0
//                    nowX = bg_on.getWidth() - slipper_pause.getWidth();
            nowStatus = ON;
        } else if (x < bg_on.getWidth() / 3) {// off 2
            nowX = 0;
            nowStatus = OFF;
        } else if (x>bg_on.getWidth()/3&&x<bg_on.getWidth()/3*2){//pause 2
//                    nowX = bg_on.getWidth() / 3;
            nowStatus = PAUSE;
        }
    }

    // 设置滑动开关的初始状态，供外部调用
    // 暂时先不需要设置nowX
    public void setNowStatus(int status) {
        switch (status) {
            case 0:// on
//                nowX = 0;
                break;
            case 2:// off
//                nowX = bg_on.getWidth() - slipper_pause.getWidth();
                break;
            case 1:// pause
//                nowX = bg_on.getWidth() / 3;
                break;
        }
    }

    // 回调接口
    public interface OnChangedListener {
        void onChanged(WiperSwitch wiperSwitch, int status);
    }

    public void setOnChangedListener(OnChangedListener listener) {
        this.listener = listener;
    }
}

File: module_views/src/main/java/com/monk/moduleviews/views/WiperSwitch.kt
package com.monk.moduleviews.views

import android.content.Context
import android.graphics.*
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.View
import android.view.View.OnTouchListener
import com.monk.commonutils.LogUtil
import com.monk.moduleviews.R

/**
 * @author MonkHank
 * @date 2018-05-10.
 */
class WiperSwitch : View, OnTouchListener {
    private var bg_on: Bitmap? = null
    private var bg_pause: Bitmap? = null
    private var bg_off: Bitmap? = null
    private var slipper_pause: Bitmap? = null
    private var slipper_white: Bitmap? = null
    // 当前滑块x
    /**
     * 0：bg_on.getWidth() - slipper_pause.getWidth()
     * 1：0
     * 2：bg_on.getWidth() / 3
     */
    private var nowX = 0

    // 0:on  1:pause  2:off
    private var nowStatus = -1
    private val ON = 0
    private val PAUSE = 1
    private val OFF = 2
    private var matrixs: Matrix? = null
    private var paint: Paint? = null
    private var listener: OnChangedListener? = null

    constructor(context: Context?) : super(context) {
        init()
    }

    constructor(context: Context?, attrs: AttributeSet?) : super(context, attrs) {
        init()
    }

    private fun init() {
        bg_off = BitmapFactory.decodeResource(resources, R.drawable.slide_switch_off)
        bg_on = BitmapFactory.decodeResource(resources, R.drawable.slide_switch_on)
        bg_pause = BitmapFactory.decodeResource(resources, R.drawable.slide_switch_stop)
        slipper_pause = BitmapFactory.decodeResource(resources, R.drawable.slide_dot)
        slipper_white = BitmapFactory.decodeResource(resources, R.drawable.slide_switch_block_30x90)
        setOnTouchListener(this)
        matrixs = Matrix()
        paint = Paint()
        nowX = bg_on?.width!! / 3
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        if (nowX < bg_on!!.width / 3) {
            canvas.drawBitmap(bg_off!!, matrix!!, paint) //画出关闭时候的背景
        } else if (nowX > bg_on!!.width / 3 * 2) {
            canvas.drawBitmap(bg_on!!, matrix!!, paint) //画出打开时候的背景
        } else {
            if (nowStatus == ON) {
                canvas.drawBitmap(bg_on!!, matrix!!, paint)
            } else if (nowStatus == OFF) {
                canvas.drawBitmap(bg_off!!, matrix!!, paint)
            } else if (nowStatus == PAUSE) {
                canvas.drawBitmap(bg_pause!!, matrix!!, paint) //画出暂停时候的背景
            }
        }
        // 对滑块滑动进行异常处理，不能让滑块出界
        if (nowX < 0) {
            nowX = 0
        } else if (nowX > bg_on!!.width - slipper_pause!!.width) {
            nowX = bg_on!!.width - slipper_pause!!.width
        }
        // 画出滑块
        LogUtil.v("WiperSwitch_onDraw", "nowX:" + nowX + "\tbg_on.getWidth():" + bg_on!!.width
                + "\tslipper_pause.getWidth():" + slipper_pause!!.width)
        when (nowStatus) {
            -1 -> {
                canvas.drawBitmap(bg_pause!!, matrix!!, paint) //画出暂停时候的背景
                canvas.drawBitmap(slipper_white!!, (bg_off!!.width / 3).toFloat(), 0f, paint)
            }
            ON, OFF -> canvas.drawBitmap(slipper_white!!, nowX.toFloat(), 0f, paint)
            PAUSE -> canvas.drawBitmap(slipper_pause!!, (bg_off!!.width / 3).toFloat(), 0f, paint)
        }
        if (action == MotionEvent.ACTION_UP) {
            if (nowX < bg_on!!.width / 3) {
                nowX += 10
                if (nowX > bg_on!!.width / 3) nowX = bg_on!!.width / 3
                postInvalidateDelayed(10)
            } else if (nowX > bg_on!!.width / 3) {
                nowX -= 10
                if (nowX < bg_on!!.width / 3) nowX = bg_on!!.width / 3
                postInvalidateDelayed(10)
            }
        }
    }

    private var action = 0
    override fun onTouch(v: View, event: MotionEvent): Boolean {
        parent.requestDisallowInterceptTouchEvent(true)
        action = event.action
        val x = event.x
        val y = event.y
        when (action) {
            MotionEvent.ACTION_DOWN -> {
                nowX = if (x > bg_on!!.width || y > bg_on!!.height) {
                    return false
                } else {
                    x.toInt()
                }
                LogUtil.e("ACTION_UP", "downX:$x")
                obtainStatus(x)
            }
            MotionEvent.ACTION_MOVE -> {
                nowX = x.toInt()
                LogUtil.i("ACTION_UP", "moveX:$x")
                obtainStatus(x)
            }
            MotionEvent.ACTION_UP -> {
                LogUtil.v("ACTION_UP", "nowStatus:$nowStatus")
                if (listener != null) listener!!.onChanged(this, nowStatus)
            }
        }
        invalidate()
        return true
    }

    private fun obtainStatus(x: Float) {
        if (x >= bg_on!!.width / 3 * 2) { // on 0
//                    nowX = bg_on.getWidth() - slipper_pause.getWidth();
            nowStatus = ON
        } else if (x < bg_on!!.width / 3) { // off 2
            nowX = 0
            nowStatus = OFF
        } else if (x > bg_on!!.width / 3 && x < bg_on!!.width / 3 * 2) { //pause 2
//                    nowX = bg_on.getWidth() / 3;
            nowStatus = PAUSE
        }
    }

    // 设置滑动开关的初始状态，供外部调用
    // 暂时先不需要设置nowX
    fun setNowStatus(status: Int) {
        when (status) {
            0 -> {
            }
            2 -> {
            }
            1 -> {
            }
        }
    }

    // 回调接口
    interface OnChangedListener {
        fun onChanged(wiperSwitch: WiperSwitch?, status: Int)
    }

    fun setOnChangedListener(listener: OnChangedListener?) {
        this.listener = listener
    }
}